### ISSUE TITLE
‘optimization’ affects buildtype

### ISSUE THREAD

**Issue by kloetzl** [mesonbuild/meson#5814]
Here is some behavior which I find very confusing.

    project('helloworld', 'c', version : '0.2', license : 'GPL3',
        default_options : [
            'buildtype=release',
            'optimization=2',
            'b_ndebug=if-release'
        ]
    )

Setting `optimization` to 2 disables the `release` buildtype. Even setting it explicitly doesn't help.

    % meson setup --buildtype=release dir
    …
    Option buildtype is: custom [default: release]

This also breaks the `if-release` feature. I think the above configuration simply should result in a `release` buildtype with `-O2` instead of `-O3`.

**Comment by jpakkane** [comment]
The semantics of buildtype are roughly "if release, set options x, y and z to values a, b and c". You should either set all individual options yourself or just set the build type.

This is due to backwards compatibility. If we were to design this from scratch, buildtype would not be a standalone option, just a shortcut to set the other options.

**Comment by kloetzl** [comment]
Thank you for looking into this. I think there are multiple facets to the issue.

> The semantics of buildtype are roughly "if release, set options x, y and z to values a, b and c".
> 
> If we were to design this from scratch, buildtype would not be a standalone option, just a shortcut to set the other options.

Well, the behavior you are describing is a shortcut, just as it should be in my opinion. If release use `-O3`, unless overridden. Above, I want to change `-O3` to `-O2`.

Now the weird thing is that setting `optimization` changes the buildtype. I don't see why it should. Furthermore, note that all these options are *default* options. Thus configuring the build via the commandline should take precedence. Interestingly, `--buildtype=release` doesn't work, but `-D optimization=3` does.

**Comment by chuling** [comment]
Agreed, I just want to change the optimization level from O3 to O2, but the buildtype 
will be changed to custom, not keep release.

I think that maybe other opts should affect buildtype, but optimization should not, it's very special.

**Comment by jpakkane** [comment]
As said, that is due to backwards compatibility. But if at all possible do not depend on buildtype, but instead look at specific settings instead.

**Comment by bonzini** [comment]
This works these days

```
  buildtype                      release                          [plain, debug, debugoptimized,   Build type to use                                            
                                                                   release, minsize, custom]                                                                    
  optimization                   2                                [plain, 0, g, 1, 2, 3, s]        Optimization level                                           
```
